manager/allocator/network/port/port.go

```diff
@@ -0,0 +1,470 @@
+	// ports allocated so far regardless of whether it was user defined or not.
+	masterPortEnd uint32 = 65535
+)
+
+// Allocator is the interface for the port allocator, which chooses and keeps
+// track of assigned and available port resources.
+type Allocator interface {
+	Restore([]*api.Endpoint)
+	Allocate(*api.Endpoint, *api.EndpointSpec) (Proposal, error)
+	Deallocate(*api.Endpoint) Proposal
+}
+
+// Allocator is an allocator component that manages the state of which
+// ports (and which protocols) are in use in the the cluster.
+//
+// The Allocator totally owns the Endpoint.Ports field. It provides a slice
+// of PortConfigs which should be written directly without modification to the
+// the Endpoint.Ports field. Nothing should be modified about Endpoint.Ports
+// outside of the port allocator. If this rule is followed, the the port
+// allocator doesn't have to check for cases like two services having the same
+// port allocated; we created all of the fields, and we can know they are
+// consistent.
+type allocator struct {
+	// ports maps the ports in use. essentially, using map as a set
+	ports map[port]struct{}
+}
+
+// port is the minimal representation for the Allocator of a single port.
+// being composed of 2 numeric types, this type is comparable and can be used
+// as a map key
+type port struct {
+	// protocol represented by this port space
+	protocol api.PortConfig_Protocol
+	port     uint32
+}
+
+// String is a quick implementation of the Stringer interface for the port
+// object so that we can pass it to string format calls with just %v
+func (p port) String() string {
+	return fmt.Sprintf("%v/%v", p.port, p.protocol)
+}
+
+// Proposal is the return value of Allocate and DeallocateEndpoint,
+// which contains the data needed to affirmatively alter the state of the
+// Allocator. It exists so that Allocator changes can be abandoned
+// without complex rollback logic. To commit changes to the Allocator, call
+// the proposal's Commit function
+type Proposal interface {
+	// Ports returns the port assignments generated by this proposal
+	Ports() []*api.PortConfig
+
+	// Commit commits the result of the Allocator transaction and alters
+	// the state of the port allocator. Part of the contract of the
+	// Allocator is that it should always return a valid proposal, and so
+	// Commit will always succeed and does not need to return an error.
+	Commit()
+
+	// IsNoop returns true if the Proposal doesn't actually modify the net
+	// allocator state. IsNoop does not mean the ports for an endpoint haven't
+	// changed. It only means that the publish ports marked in use by the
+	// allocator haven't changed.
+	IsNoop() bool
+}
+
+type proposal struct {
+	pa         *allocator
+	ports      []*api.PortConfig
+	allocate   map[port]struct{}
+	deallocate map[port]struct{}
+}
+
+func (prop *proposal) Ports() []*api.PortConfig {
+	if prop.ports == nil {
+		return []*api.PortConfig{}
```

Comment:
Why not just return nil ?

Reply:
nil is evil

Reply:
care to elaborate ? I would love to hear more on this

Reply:
nil sneaks up on you when you're not looking, nil checks make everything uglier, etc etc. Go treats some things that are nil as things and some things that are nil as nil. It's just more hygienic to avoid the use of nil unless a value is intended to be discarded.